cmd/coder/main.go

```diff
@@ -31,16 +30,19 @@ func main() {
 		clog.Log(clog.Fatal(fmt.Sprintf("set output to raw: %s", err)))
 		os.Exit(1)
 	}
-	defer func() {
+	restoreTerminal := func() {
 		// Best effort. Would result in broken terminal on window but nothing we can do about it.
 		_ = xterminal.Restore(os.Stdout.Fd(), stdoutState)
-	}()
+	}
 
 	app := cmd.Make()
 	app.Version = fmt.Sprintf("%s %s %s/%s", version.Version, runtime.Version(), runtime.GOOS, runtime.GOARCH)
 
 	if err := app.ExecuteContext(ctx); err != nil {
 		clog.Log(err)
+		restoreTerminal()
 		os.Exit(1)
 	}
+	restoreTerminal()
+	cancel()
```

Comment:
I was very skeptical of `gocritic` but the fact that it caught this bug was enough to convince me.